File: mobile/lib/domain/engines/rule_switch_engine.dart
import 'dart:math';
import '../models/game_stimulus.dart';

enum RuleType {
  color,
  shape,
  negation, // "NOT red" or "NOT circle"
}

class RuleSwitchRound {
  final int roundNumber;
  final RuleType ruleType;
  final String ruleDescription;
  final String targetCriterion;
  final bool isSwitchTrial;
  final List<GameStimulus> choices;
  final int correctChoiceIndex;

  RuleSwitchRound({
    required this.roundNumber,
    required this.ruleType,
    required this.ruleDescription,
    required this.targetCriterion,
    required this.isSwitchTrial,
    required this.choices,
    required this.correctChoiceIndex,
  });
}

class RuleSwitchEngine {
  final Random _random = Random();
  RuleType _currentRule = RuleType.color;
  int _trialsUnderCurrentRule = 0;

  RuleSwitchRound generateRound({required int roundNumber, required double level}) {
    // Switch rule every 2 to 3 trials to measure task-switch cost
    final bool shouldSwitch = _trialsUnderCurrentRule >= 2 && (_random.nextBool() || _trialsUnderCurrentRule >= 3);
    final bool isSwitch = shouldSwitch || roundNumber == 1;

    if (isSwitch && roundNumber > 1) {
      final available = RuleType.values.where((r) => r != _currentRule).toList();
      // Only include negation rule at difficulty level >= 3.0
      if (level < 3.0) {
        available.remove(RuleType.negation);
      }
      _currentRule = available[_random.nextInt(available.length)];
      _trialsUnderCurrentRule = 1;
    } else {
      _trialsUnderCurrentRule++;
    }

    final shapes = TargetShape.values;
    final colors = TargetColor.values;

    String ruleDesc;
    String targetCriterion;
    int correctIndex = 0;
    List<GameStimulus> choices = [];

    if (_currentRule == RuleType.color) {
      final targetColor = colors[_random.nextInt(colors.length)];
      ruleDesc = 'RULE: MATCH COLOR';
      targetCriterion = 'Tap ${targetColor.name.toUpperCase()}';

      // 4 choices
      final targetShape = shapes[_random.nextInt(shapes.length)];
      final correctStim = GameStimulus(
        id: 'rs_${roundNumber}_0',
        shape: targetShape,
        color: targetColor,
        isTarget: true,
        accessibleLabel: '${targetColor.name} ${targetShape.name}',
      );

      choices.add(correctStim);

      // 3 wrong colors
      final wrongColors = colors.where((c) => c != targetColor).toList()..shuffle(_random);
      for (int i = 0; i < 3; i++) {
        choices.add(GameStimulus(
          id: 'rs_${roundNumber}_${i + 1}',
          shape: shapes[_random.nextInt(shapes.length)],
          color: wrongColors[i],
          isTarget: false,
          accessibleLabel: wrongColors[i].name,
        ));
      }
    } else if (_currentRule == RuleType.shape) {
      final targetShape = shapes[_random.nextInt(shapes.length)];
      ruleDesc = 'RULE: MATCH SHAPE';
      targetCriterion = 'Tap ${targetShape.name.toUpperCase()}';

      final targetColor = colors[_random.nextInt(colors.length)];
      final correctStim = GameStimulus(
        id: 'rs_${roundNumber}_0',
        shape: targetShape,
        color: targetColor,
        isTarget: true,
        accessibleLabel: '${targetColor.name} ${targetShape.name}',
      );

      choices.add(correctStim);

      // 3 wrong shapes
      final wrongShapes = shapes.where((s) => s != targetShape).toList()..shuffle(_random);
      for (int i = 0; i < 3; i++) {
        choices.add(GameStimulus(
          id: 'rs_${roundNumber}_${i + 1}',
          shape: wrongShapes[i],
          color: colors[_random.nextInt(colors.length)],
          isTarget: false,
          accessibleLabel: wrongShapes[i].name,
        ));
      }
    } else {
      // Negation rule (e.g. Tap anything EXCEPT RED)
      final forbiddenColor = colors[_random.nextInt(colors.length)];
      ruleDesc = 'RULE: INHIBITION';
      targetCriterion = 'Tap NOT ${forbiddenColor.name.toUpperCase()}';

      // 3 choices have the forbidden color, 1 does not (the correct one)
      final safeColor = colors.firstWhere((c) => c != forbiddenColor);
      final correctStim = GameStimulus(
        id: 'rs_${roundNumber}_0',
        shape: shapes[_random.nextInt(shapes.length)],
        color: safeColor,
        isTarget: true,
        accessibleLabel: safeColor.name,
      );
      choices.add(correctStim);

      for (int i = 0; i < 3; i++) {
        choices.add(GameStimulus(
          id: 'rs_${roundNumber}_${i + 1}',
          shape: shapes[_random.nextInt(shapes.length)],
          color: forbiddenColor,
          isTarget: false,
          accessibleLabel: forbiddenColor.name,
        ));
      }
    }

    choices.shuffle(_random);
    correctIndex = choices.indexWhere((c) => c.isTarget);

    return RuleSwitchRound(
      roundNumber: roundNumber,
      ruleType: _currentRule,
      ruleDescription: ruleDesc,
      targetCriterion: targetCriterion,
      isSwitchTrial: isSwitch,
      choices: choices,
      correctChoiceIndex: correctIndex,
    );
  }
}

File: web/src/engines/ruleSwitchEngine.ts
import type { TargetShape, TargetColor, GameStimulus } from '../types/game';

export type RuleType = 'color' | 'shape' | 'negation';

export interface RuleSwitchRound {
  roundNumber: number;
  ruleType: RuleType;
  ruleDescription: string;
  targetCriterion: string;
  isSwitchTrial: boolean;
  choices: GameStimulus[];
  correctChoiceIndex: number;
}

const SHAPES: TargetShape[] = ['circle', 'triangle', 'square', 'diamond', 'star'];
const COLORS: TargetColor[] = ['blue', 'red', 'green', 'yellow', 'purple', 'orange'];

export class RuleSwitchEngine {
  private currentRule: RuleType = 'color';
  private trialsUnderRule: number = 0;

  generateRound(roundNumber: number, level: number): RuleSwitchRound {
    const shouldSwitch = this.trialsUnderRule >= 2 && (Math.random() > 0.5 || this.trialsUnderRule >= 3);
    const isSwitch = shouldSwitch || roundNumber === 1;

    if (isSwitch && roundNumber > 1) {
      const available: RuleType[] = ['color', 'shape'];
      if (level >= 3.0) available.push('negation');
      const filtered = available.filter(r => r !== this.currentRule);
      this.currentRule = filtered[Math.floor(Math.random() * filtered.length)];
      this.trialsUnderRule = 1;
    } else {
      this.trialsUnderRule++;
    }

    let ruleDescription = '';
    let targetCriterion = '';
    const choices: GameStimulus[] = [];

    if (this.currentRule === 'color') {
      const targetColor = COLORS[Math.floor(Math.random() * COLORS.length)];
      ruleDescription = 'RULE: MATCH COLOR';
      targetCriterion = `Tap ${targetColor.toUpperCase()}`;

      choices.push({
        id: `rs_${roundNumber}_0`,
        shape: SHAPES[Math.floor(Math.random() * SHAPES.length)],
        color: targetColor,
        isTarget: true,
        accessibleLabel: targetColor,
      });

      const wrong = COLORS.filter(c => c !== targetColor).sort(() => Math.random() - 0.5);
      for (let i = 0; i < 3; i++) {
        choices.push({
          id: `rs_${roundNumber}_${i + 1}`,
          shape: SHAPES[Math.floor(Math.random() * SHAPES.length)],
          color: wrong[i],
          isTarget: false,
          accessibleLabel: wrong[i],
        });
      }
    } else if (this.currentRule === 'shape') {
      const targetShape = SHAPES[Math.floor(Math.random() * SHAPES.length)];
      ruleDescription = 'RULE: MATCH SHAPE';
      targetCriterion = `Tap ${targetShape.toUpperCase()}`;

      choices.push({
        id: `rs_${roundNumber}_0`,
        shape: targetShape,
        color: COLORS[Math.floor(Math.random() * COLORS.length)],
        isTarget: true,
        accessibleLabel: targetShape,
      });

      const wrong = SHAPES.filter(s => s !== targetShape).sort(() => Math.random() - 0.5);
      for (let i = 0; i < 3; i++) {
        choices.push({
          id: `rs_${roundNumber}_${i + 1}`,
          shape: wrong[i],
          color: COLORS[Math.floor(Math.random() * COLORS.length)],
          isTarget: false,
          accessibleLabel: wrong[i],
        });
      }
    } else {
      const forbidden = COLORS[Math.floor(Math.random() * COLORS.length)];
      ruleDescription = 'RULE: INHIBITION';
      targetCriterion = `Tap NOT ${forbidden.toUpperCase()}`;

      const safe = COLORS.find(c => c !== forbidden) || 'blue';
      choices.push({
        id: `rs_${roundNumber}_0`,
        shape: SHAPES[Math.floor(Math.random() * SHAPES.length)],
        color: safe,
        isTarget: true,
        accessibleLabel: safe,
      });

      for (let i = 0; i < 3; i++) {
        choices.push({
          id: `rs_${roundNumber}_${i + 1}`,
          shape: SHAPES[Math.floor(Math.random() * SHAPES.length)],
          color: forbidden,
          isTarget: false,
          accessibleLabel: forbidden,
        });
      }
    }

    choices.sort(() => Math.random() - 0.5);
    const correctChoiceIndex = choices.findIndex(c => c.isTarget);

    return {
      roundNumber,
      ruleType: this.currentRule,
      ruleDescription,
      targetCriterion,
      isSwitchTrial: isSwitch,
      choices,
      correctChoiceIndex,
    };
  }
}
